### PR TITLE
jwt authentication

### LINKED ISSUE #62: jwt authentication
- [x] allow registration of JWKs providers with a custom refresh rates

- [ ] retrieve JWT claims from cached JWTs
verify uncached JWTs using JWKs
validate JWTs against sets of claim rules
cache results based on JWT expiration / other TTL

similar to [api key caching](https://github.com/apigee/apigee-remote-service-golib/blob/master/auth/verify_api_key.go#L18)

### PR DESCRIPTION
**Pull request by theganyo**
Fixes #62

### CODE REVIEW

**Comment by rockspore** [comment]
If we merge this, I suppose we need to eventually make a 2.0 tag more carefully?

File: auth/jwt/jwt.go [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+const (
+	acceptableSkew               = 10 * time.Second
+	defaultCacheTTL              = 30 * time.Minute
+	defaultCacheEvictionInterval = 10 * time.Second
+	defaultMaxCachedEntries      = 10000
+	defaultBadEntryCacheTTL      = 10 * time.Second
+	minAllowedRefreshInterval    = 10 * time.Minute
+)
+
+func NewVerifier(opts VerifierOptions) Verifier {
```
Comment (rockspore) [inline auth/jwt/jwt.go:44]: Exported functions need documentation.

File: auth/jwt/jwt.go [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+type verifier struct {
+	jwks          *jwk.AutoRefresh
+	cancelContext context.Context
+	cancelFunc    context.CancelFunc
+	providers     []Provider
+	cache         cache.ExpiringCache
+	knownBad      cache.ExpiringCache
+}
+
+func (a *verifier) Start() {
```
Comment (rockspore) [inline auth/jwt/jwt.go:91]: Same here.

File: auth/jwt/jwt.go [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+	for i := range a.providers {
+		p := a.providers[i]
+		if _, err := a.jwks.Refresh(ctx, p.JWKSURL); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (a *verifier) Stop() {
```
Comment (rockspore) [inline auth/jwt/jwt.go:114]: Same here.

File: auth/jwt/jwt.go [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+			if minRefresh < minAllowedRefreshInterval {
+				minRefresh = minAllowedRefreshInterval
+			}
+			options = append(options, jwk.WithMinRefreshInterval(minRefresh))
+		}
+		a.jwks.Configure(provider.JWKSURL, options...)
+	}
+}
+
+func (a *verifier) FetchJWKs(provider Provider) (jwk.Set, error) {
```
Comment (rockspore) [inline auth/jwt/jwt.go:147]: Same here.

**Review by rockspore — APPROVED** [review]

**Comment by theganyo** [comment]
Re: merge... Technically, yes, although there is no behavioral change.

**Pull request merged**